src/cargo/core/compiler/fingerprint/rustdoc.rs

```diff
@@ -0,0 +1,150 @@
+use std::path::Path;
+
+use anyhow::Context as _;
+use cargo_util::paths;
+use serde::Deserialize;
+use serde::Serialize;
+
+use crate::CargoResult;
+use crate::core::compiler::BuildRunner;
+use crate::core::compiler::CompileKind;
+
+/// Structure used to deal with Rustdoc fingerprinting
+///
+/// This is important because the `.js`/`.html` & `.css` files
+/// that are generated by Rustc don't have any versioning yet
+/// (see <https://github.com/rust-lang/cargo/issues/8461>).
+/// Therefore, we can end up with weird bugs and behaviours
+/// if we mix different versions of these files.
+///
+/// We need to make sure that if there were any previous docs already compiled,
+/// they were compiled with the same Rustc version that we're currently using.
+/// Otherwise we must remove the `doc/` folder and compile again forcing a rebuild.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct RustDocFingerprint {
+    /// `rustc -vV` verbose version output.
+    pub rustc_vv: String,
+}
+
+impl RustDocFingerprint {
+    /// Checks whether the latest version of rustc used to compile this workspace's docs
+    /// was the same as the one is currently being used in this `cargo doc` call.
+    ///
+    /// In case it's not,
+    /// it takes care of removing the `<build-dir>/doc/` folder
+    /// as well as overwriting the rustdoc fingerprint info.
+    /// This is to guarantee that we won't end up with mixed versions of the `js/html/css` files
+    /// which `rustdoc` autogenerates without any versioning.
+    ///
+    /// Each requested target platform maintains its own fingerprint file.
+    /// That is, if you run `cargo doc` and then `cargo doc --target wasm32-wasip1`,
+    /// you will have two separate fingerprint files:
+    ///
+    /// * `<build-dir>/.rustdoc_fingerprint.json` for host
+    /// * `<build-dir>/wasm32-wasip1/.rustdoc_fingerprint.json`
+    pub fn check_rustdoc_fingerprint(build_runner: &BuildRunner<'_, '_>) -> CargoResult<()> {
+        if build_runner
+            .bcx
+            .gctx
+            .cli_unstable()
+            .skip_rustdoc_fingerprint
+        {
+            return Ok(());
+        }
+        let new_fingerprint = RustDocFingerprint {
+            rustc_vv: build_runner.bcx.rustc().verbose_version.clone(),
+        };
+
+        for kind in &build_runner.bcx.build_config.requested_kinds {
+            check_fingerprint(build_runner, &new_fingerprint, *kind)?;
+        }
+
+        Ok(())
+    }
+}
+
+/// Checks rustdoc fingerprint file for a given [`CompileKind`].
+fn check_fingerprint(
```

Comment:
Can you split this function out in its own commit to make the before/after comparison easier?